### PR TITLE
Update docstrings. Add download function. Add more auth methods. Clean up code.

### PR DESCRIPTION
**Pull request by Jayy001**

### CODE REVIEW

File: pyfactorybridge/__init__.py [hunk 1]
```diff
@@ -121,54 +120,125 @@ def __request(
...
+        permissions: str = "Administrator",
+    ):
+        """Renew the authentication token based on the provided method."""
+
+        if permissions not in [
+            "Administrator",
+            "Client",
+            "APIToken",
+            "NotAuthenticated",
+        ]:
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:136]: Maybe this could be handled a bit more elegantly using enums instead

File: pyfactorybridge/__init__.py [hunk 2]
```diff
@@ -121,54 +120,125 @@ def __request(
...
+            logging.error("No password provided, trying passwordless login.")
+            self.auth = self.__auth_from_passwordless()
+
+            if self.auth:
+                logging.info("Passwordless login successful. Please claim the server.")
+
+            else:
+                logging.error("Passwordless login failed. Using no authentication.")
+
+        logging.info(f"API initialized with auth token: {str(self.auth)}")
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:157]: Replace with
`logging.info(f"API initialized with auth token: {self.auth}")`
(f-strings implicitly call `__str__` already)

File: pyfactorybridge/__init__.py [hunk 3]
```diff
@@ -121,54 +120,125 @@ def __request(
...
             return BearerAuth(token)
 
-    def authorise_api_token(self, Token) -> BearerAuth:
+    def __auth_from_passwordless(self) -> BearerAuth | None:
+        response_data = self.__request(
+            function="PasswordlessLogin",
+            properties={"minimumPrivilegeLevel": "InitialAdmin"},
+        )
+
+        if (token := response_data.get("authenticationToken")) is not None:
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:179]: Is `is not None` required here?

File: pyfactorybridge/__init__.py [hunk 4]
```diff
@@ -177,45 +247,56 @@ def set_admin_password(self, AdminPassword, AuthenticationToken) -> dict[str, st
...
 
-    def set_auto_load_session_name(self, SessionName) -> dict[str, str]:
+    def set_auto_load_session_name(self, SessionName: str) -> dict[str, str]:
+        """Sets the auto load session name. Returns nothing."""
         return self.__request(
             function="SetAutoLoadSessionName", properties={"SessionName": SessionName}
         )
 
-    def run_command(self, Command) -> dict[str, str]:
+    def run_command(self, Command: str) -> dict[str, str]:
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:256]: Missing doc-string

File: pyfactorybridge/__init__.py [hunk 5]
```diff
@@ -231,31 +312,44 @@ def upload_save_game(
...
-                properties={
-                    "SaveName": SaveName
-                    or save_file_path_obj.stem,  # Save file name without file extension
-                    "LoadSaveGame": LoadSaveGame,
-                    "EnableAdvancedGameSettings": EnableAdvancedGameSettings,
-                },
-                multiparts=multiparts,
-            )
+        except (FileNotFoundError, PermissionError, OSError):
+            raise ServerError("Cannot read file path")
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:340]: I dislike "ServerError" here, but is fine for now I'd say

File: pyfactorybridge/__init__.py [hunk 6]
```diff
@@ -231,31 +312,44 @@ def upload_save_game(
...
 
-    def download_save_game(self, SaveName) -> None:
-        raise NotImplementedError
+    def download_save_game(self, SaveName: str, Path: str) -> None:
```
Comment (janerjak) [inline pyfactorybridge/__init__.py:342]: Is path an API-argument? Else it should probably be lower case

**Review by janerjak — COMMENTED** [review]
Small nitpicks, but looks really good 👍

File: pyfactorybridge/__init__.py (hunk 3 above)
Comment (Jayy001) [reply pyfactorybridge/__init__.py:179]: No, good catch

File: pyfactorybridge/__init__.py (hunk 5 above)
Comment (Jayy001) [reply pyfactorybridge/__init__.py:340]: Agreed, I'll change that

File: pyfactorybridge/__init__.py (hunk 1 above)
Comment (Jayy001) [reply pyfactorybridge/__init__.py:136]: I don't think an enum would be useful here

**Pull request merged**